[영화 티켓 예매 1, 2단계] 채넛 미션 제출합니다

yrsel opened the pull request:
# 셀프 체크리스트

 <!--
 모든 항목을 점검한 후 리뷰를 요청하면, 보다 의미 있는 피드백을 받을 수 있습니다.
 -->

- [x] 프로그램이 정상적으로 작동하는가?
- [x] 모든 테스트가 통과하는가?
- [ ] 이전에 받은 피드백을 모두 반영하였는가?
- [x] 코딩 스타일 가이드를 준수하였는가?
    - [x] IDE 코드 자동 정렬을 적용하였는가?
    - [x] 린트 검사를 통과하였는가?
- [x] 프로그래밍 요구 사항을 준수하였는가?
- [x] `README.md`에 기능 목록을 정리하고 명확히 기술하였는가?

## 코드 리뷰 커뮤니케이션

안녕하세요, 코니 
영화 티켓 예매 리뷰이 채넛입니다. 🙇‍♂️
이번 미션동안 잘 부탁드립니다 ! 

# 어떤 부분에 집중하여 리뷰해야 할까요?

미션을 진행하면서 궁금했던 부분에 대해서 질문을 남겼습니다.

🙋‍♂️ 질문 1. 날짜 및 시간 포맷팅 책임의 적절한 위치
이번 미션을 진행하면서 날짜와 시간 포맷팅이 필요한 상황이 있었습니다.
View에서는 String 형태로 데이터를 다루고, 도메인 계층에서는 LocalDate나 LocalTime과 같은 객체를 사용하여 비교를 진행하다 보니, Activity나 도메인 계층에서 포맷 변환 로직이 필요했습니다.
단순히 Util 클래스를 만들어 처리하는 것보다는, 책임에 맞는 객체에 포맷팅 역할을 위임하고 싶은데, 이 경우 포맷팅 책임을 어디에 두는 것이 가장 적절할지 고민이 됩니다.

🙋‍♂️ 질문 2. Intent를 통해 data class 객체를 전달할 때 getParcelableExtra를 사용했습니다.
하지만 하위 버전(Android 12 이하)에서는 해당 메서드가 deprecated되어 분기 처리가 필요했고, 결국 deprecated된 방식도 함께 사용할 수밖에 없었습니다.
하위 버전까지 고려했을 때 이 문제를 깔끔하게 처리할 수 있는 방법이나, 권장되는 처리 방식이 있다면 알고 싶습니다.

🙋‍♂️ 질문 3. `BookingResult`를 immutable하게 유지하고 싶었지만, 일부 값들이 외부에서 변경되는 상황이 발생해 backing property를 사용해 값을 캡슐화했습니다. 이런 방식이 최선이었을지 고민되고, 더 나은 구조나 대안이 있다면 알고 싶습니다.

🙋‍♂️ 질문 4. 외부에서 값을 주입하기 위해 `lateinit 프로퍼티`를 사용하고 있는데, 이 방식이 적절한지도 궁금합니다.
객체 생성 이후 특정 시점에 값을 설정해야 해서 `lateinit`을 사용했는데, 이로 인해 nullable 처리나 초기화 시점 관리가 모호해지는 느낌도 있습니다. 하지만, null값을 넣어주는 것보다는 lateinit 프로퍼티를 사용하는 것이 더 낫다고 생각했습니다.
더 나은 구조로 개선할 수 있는 방법이 있을지 궁금합니다.

yrsel committed:
docs: README 기능 목록 작성

yrsel committed:
feat: 화면에 영화 리스트 구현

gahyunkim committed:
feat: 예매 버튼을 선택 시 기본 정보를 담은 화면 구현

yrsel committed:
feat: 인텐트 테스트 의존성 추가

yrsel committed:
test: 메인 액티비티 뷰 UI 테스트

gahyunkim committed:
feat: 영화 예매 화면 구현

yrsel committed:
feat: 영화 예매 도메인 구현

gahyunkim committed:
feat: 영화 예매 인원 선택 및 다이알로그 구현

yrsel committed:
fix: 오작동 코드 수정

yrsel committed:
feat: 다이얼로그 외부 클릭 방지 구현

yrsel committed:
feat: 영화 예매 완료 안내페이지 구현

gahyunkim committed:
feat: 화면 회전 시 입력 정보 유지 및 세부 기능 구현

gahyunkim committed:
test: UiTest 구현

yrsel committed:
style: 소스 코드 함수로 분리 및 정리

yrsel committed:
feat: 예약현황 직접 처리하도록 구현

gahyunkim committed:
test: UiTest 구현

- BookingActivityTest, BookingCompleteActivityTest 구현

gahyunkim committed:
feat: Movie 예외 처리 구현

gahyunkim committed:
style: MainActivity 구조화

gahyunkim committed:
feat: BookingResult 예외처리 구현

gahyunkim committed:
design: BookingCompleteActivity 스크롤뷰 구현

yrsel committed:
test: Movie 예외 테스트 구현

yrsel committed:
test: Booking 예외 테스트 구현

yrsel committed:
test: BookingResult 예외 테스트 구현

yrsel committed:
style: 에러메시지 상수화

yrsel committed:
style: 주석 제거

yrsel committed:
style: 코드 위치 수정

lee-ji-hoon commented on app/src/androidTest/java/woowacourse/movie/BookingActivityTest.kt:
```diff
@@ -0,0 +1,296 @@
+        onView(withId(R.id.tv_people_count))
+            .check(matches(withText("1")))
+
+        onView(withId(R.id.btn_minus))
+            .perform(click())
+
+        onView(withId(R.id.tv_people_count))
+            .check(matches(withText("0")))
+    }
+
+    @Test
+    fun `확인_버튼을_누르면_다이알로그가_뜬다`() {
```
UI Test도 결국 `Test` 이기에 UnitTest 하던때처럼 given / when / then 을 할수도 있습니다.

> 저는 저렇게 어떠한 `가정` 이 필요하게 되면 given / when / then을 사용하긴 합니다

yrsel replied:
관련 커밋 : e36efb7f9c28673a6291445ca6e548141803f8a4

lee-ji-hoon commented on app/src/androidTest/java/woowacourse/movie/BookingActivityTest.kt:
```diff
@@ -0,0 +1,296 @@
+    }
+
+    @Test
+    fun `예매_완료_버튼을_누르면_다음_화면으로_데이터를_가지고_넘어간다`() {
```
UI Test를 할 때는 현재 화면에 어떠한 상태가 잘 보여지는지 아니면 어떠한 시나리오를 했을 때 화면이 잘 표기가 되는지. 이런 것들을 메인으로 테스트를 한다고 생각을 해보면 이 테스트가 꼭 필요할까요? 

고민이 될만한 포인트인데요. 위 제 생각처럼 채넛이 생각했을 때 어느 범위까지 테스트가 필요한지 직접 판단을 해보면 좋을 것 같으니 한 번 의견을 말씀해주시면 좋을 것 같네요~!

yrsel replied:
Espresso 에서 intended 라는 메서드를 제공해주니 당연히 UI 테스트 범위라고 생각하고 코드를 작성했었습니다.
말씀해주신 내용을 기반으로 다시 생각해보니 UI test는 화면에 데이터가 잘 표시되는지, 화면이 잘 그려지는지에 대한 테스트가 주요 역할이라고 생각됩니다.

이런 관점에서 데이터를 가지고 잘 넘어가는 지, 넘어간 데이터가 잘 표시되었는 것은 어떻게 보면 Intent key값이 잘 들어갔는 지 테스트하는 것 같아서 UI 테스트와는 조금 다른 결 인 거 같기도 합니다. ☺️

lee-ji-hoon commented on app/src/androidTest/java/woowacourse/movie/BookingActivityTest.kt:
```diff
@@ -0,0 +1,296 @@
+import woowacourse.movie.model.Movie
+import java.time.LocalDate
+
+class BookingActivityTest {
```
UI Test 잘 구현을 해주셨습니다.

그럼 espresso를 사용한 김에 화면 회전간에도 데이터가 유지가 되는지도 확인하는 테스트도 구현을 해보면 어떨까요?

화면 회전에 대한 action에서도 ui의 state가 유지가 되고 표시가 되는지를 확인해보는 것입니다!

yrsel replied:
관련 커밋 : 9a7b33e3965f027f65dd1467f3c43dbb90364369

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+        setContentView(R.layout.activity_booking)
+        setUpUi()
+
+        val movieData = movieOrNull() ?: return
```
만약 data가 null이라서 return을 하게 되면 어떤 화면이 보이게 될까요?

yrsel replied:
만약, return 전에 아무런 로직도 수행하지 않았다면 빈 화면이 보일테고, setContentView 호출 이후라면 기본적인 view 계층 구조는 그려지고 데이터는 없는 상태가 될 것입니다.(view에 기본값이 들어있지 않은 경우) 

일단 현재 미션에서는 더미 영화를 넣어주도록 진행했습니다.
운영중인 서비스라면 잘못되었음을 알려주는 데이터들을 넣어주고 예매 불가능하게 만들거나 잘못된 상황을 알리는 다이얼로그를 띄우도록 해 이전 화면으로 돌려보내는 방식으로 진행해야할 것 같습니다.

아직 `BookingActivity`, `BookingCompleteActivity` 에서의 data null 일때의 로직은 수정하지 못했습니다. 🥲

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+        setUpReserveConfirm()
+    }
+
+    private fun setUpUi() {
```
`setUpUi` 라고 하니까 UI에 대한 세팅을 할 것 같이 생겼는데 막상 들어가니 inset 에 대한 조절만 하고 있네요. 이 함수가 명확한 일을 하게끔 해주면 어떨까요?

yrsel replied:
안드로이드 버전에 따라 액션바를 설정해주고 알맞은 위치에 배치하도록 조정해주는 역할을 하고 있어 수정했습니다.

관련 커밋 : aa8f1023e8e90d18d1ec418fde83e9b8a17dbb7a

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+        } else {
+            intent.getParcelableExtra("movieData")
+        }
+    }
```
> 🙋‍♂️ 질문 2. Intent를 통해 data class 객체를 전달할 때 getParcelableExtra를 사용했습니다.
하지만 하위 버전(Android 12 이하)에서는 해당 메서드가 deprecated 되어 분기 처리가 필요했고, 결국 deprecated 된 방식도 함께 사용할 수밖에 없었습니다.
하위 버전까지 고려했을 때 이 문제를 깔끔하게 처리할 수 있는 방법이나, 권장되는 처리 방식이 있다면 알고 싶습니다.

`deprecated` 된 방식이어도 버전 대응을 위해서는 어쩔 수 없이 사용을 하게 되는 상황들이 오는데요.

그런 상황에서 가장 베스트는 `Compat` 으로 모든 버전을 대응해 주는 것입니다. 이렇게 하면 `Deprecated` 된 방식을 직접적으로 사용을 하지 않고 내부적으로 감출 수 있기 때문이죠. 그렇기에 지금 구현해 두신 것도 `Compat`처럼 만들어보면 어떨까 싶네요!

yrsel replied:
직접 클래스를 생성해서 내부로 감출 수 있도록 하였습니다. 

관련 커밋 : 08da12ecae9258d2159dfdaadc4c2e3e682dc809

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+
+        outState.putInt("HEAD_COUNT", bookingResult.headCount)
+        outState.putString("SCREENING_DATE", bookingResult.selectedDate)
+        outState.putString("SCREENING_TIME", bookingResult.selectedTime)
```
레벨 1 때 했던 것을 생각하면 이런 하드 코딩된 String을 어떻게 하기로 했었을까요 👀

yrsel replied:
관련 커밋 : 6409ac822cbfbe1681a74360d9313578ae2b9b25

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+        }
+    }
+
+    private fun formatDate(
+        date: LocalDate,
+        delimiter: Char,
+    ): String {
+        return date.format(DateTimeFormatter.ofPattern("yyyy${delimiter}M${delimiter}d"))
+    }
+
+    private fun formatTime(time: LocalTime): String {
+        return time.format(DateTimeFormatter.ofPattern("kk:mm"))
+    }
```
이런 것들은 다른 화면에서도 사용이 될테니 Util 같은 것으로 하나 빼도 좋을 것 같네요

yrsel replied:
관련 커밋 : 10bf66c32d70f61264c3223cb98f85c68884c729

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+            .show()
+    }
+
+    override fun onSupportNavigateUp(): Boolean {
+        finish()
+        return super.onSupportNavigateUp()
+    }
```
여기는 어떤 의도로 구현을 하게 된 것 일까요?

yrsel replied:
액션바에 뒤로가기 버튼이 작동되려면 해당 액티비티가 종료되어야해서 `finish()` 로 종료 하고 `AppcompatActivity`의 메서드를 호출하게 구현하였습니다.

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+class BookingActivity : AppCompatActivity() {
+    private lateinit var bookingResult: BookingResult
+
+    override fun onCreate(savedInstanceState: Bundle?) {
```
> 틀렸다는 이야기가 아닙니다. 잘 알고 계신지 확인차 남기는 것입니다.

Activity의 Lifecycle 중에 `onResume` `onStart` 도 존재하는데 대부분의 초기화를 `onCreate` 에서 해주신 모습인데요. 어떤 로직을 어떠한 Lifecycle에 적용을 하는지 채넛이 판단하고 있는 기준이 있을까요?

yrsel replied:
안드로이드 생명주기에 대해 이론적으로만 알고 있어서 큰 의미없이 `onCreate`에 넣었던 것 같습니다.

`onCreate`: 최초 1회만 호출
`onStart`: 가시성만 존재하는 상태로, 데이터를 초기화하는 로직 포함 가능할 것 같습니다.
`onResume`: 상호작용 가능한 상태라고 생각합니다.

적절한 생명주기에서 알맞은 처리가 이뤄지도록 조금 더 고민해보겠습니다 !

lee-ji-hoon replied:
그럼 추가로 `restore` 를 위한 적절한 Lifecycle도 하나 존재하는데 찾아 보시면 좋을 것 같네요!~

yrsel replied:
restore Lifecycle 내부에서 인원수에 대한 Textview를 업데이트 해줘야만 configuration change 발생 시 정상 동작하고 있어, restore 해주는 곳에도 view를 업데이트 해주게 되었습니다.
올바른 방법은 아닌 것 같은데, 어떤 방식이 존재할까요 ? 
onCreate() -> onStart() -> onRestoreInstanceState() -> onResume() 순으로 호출되는 순서는 알겠으나 해결 방법이 떠오르지 않습니다. 🥹

관련 커밋 : 7a7758b8557ba0f35561f9401dcaeaceea52b8c2

lee-ji-hoon replied:
`onRestoreInstanceState` 의 경우 `onStart` 이후에 호출이 되는 것이 맞습니다. 그래서 사실 사용하기가 마냥 편하지가 않은데요. 해당 함수의 주석을 보고 의도를 파악을 해보는 것도 좋은 경험이 될 것 같네요.

아래 주석의 의도를 이해하다보면 채넛이 어색해하는 부분이 왜 그런지 이해가 될 것 같고, 그에 따라 UI의 수정 코드는 다른 곳으로 다시 옮겨가게 될 것 같네요

> This method is called after onStart() when the activity is being re-initialized from a previously saved state, given > here in savedInstanceState. Most implementations will simply use onCreate(Bundle) to restore their state, but it is > sometimes convenient to do it here after all of the initialization has been done or to allow subclasses to decide > 
whether to use your default implementation.

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+import java.time.format.DateTimeFormatter
+
+class BookingActivity : AppCompatActivity() {
+    private lateinit var bookingResult: BookingResult
```
> 🙋‍♂️ 질문 4. 외부에서 값을 주입하기 위해 lateinit 프로퍼티를 사용하고 있는데, 이 방식이 적절한지도 궁금합니다.
객체 생성 이후 특정 시점에 값을 설정해야 해서 lateinit을 사용했는데, 이로 인해 nullable 처리나 초기화 시점 관리가 모호해지는 느낌도 있습니다. 하지만, null값을 넣어주는 것보다는 lateinit 프로퍼티를 사용하는 것이 더 낫다고 생각했습니다.
더 나은 구조로 개선할 수 있는 방법이 있을지 궁금합니다.

우선 2가지로 질문을 나눠서 답변을 드릴게요.

> 외부에서 값을 주입하기 위해 lateinit 프로퍼티를 사용하고 있는데, 이 방식이 적절한지도 궁금합니다.

우선 부적절하지는 않습니다. 단 lateinit을 썼을 때 주의해야 할 사항들이 있습니다.

```kotlin
`lateinit` 으로 선언을 했을 때 가장 조심해야 하는게 있는데요.

```kotlin
private lateinit var temp: String

fun foo() {
    /**
     * 1. temp가 초기화가 안됐다면?
     * 2. 초기화가 됐었지만 Memory Kill이 발생하게 되면 어떻게 될까? 
     */
    println(temp)
}
```

이런 상황들을 조심해야 하며 위 상황들을 어떻게 대응할 수 있을지를 고민해보시기 바랍니다.


---

> 객체 생성 이후 특정 시점에 값을 설정해야 해서 lateinit을 사용했는데, 이로 인해 nullable 처리나 초기화 시점 관리가 모호해지는 느낌도 있습니다. 하지만, null값을 넣어주는 것보다는 lateinit 프로퍼티를 사용하는 것이 더 낫다고 생각했습니다.

nullable하게 해서 null 값을 넣어주는 게 그리 이상한가?라고 고민을 해보면 저는 잘 공감은 안되긴 하네요.

조금 다른 예시긴 한데 이런 예시도 있습니다.

```kotlin
enum class BookingState {
    NONE, // NONE이 상태인게 맞나? 아무것도 안했으면 null로 명확하게 하면 되지 않나?
    BOOKING,
    CANCEL
}
```

저는 저렇게 `NONE` 을 두는 것 보다 때에 따라서는 nullable한 것이 다른 개발자들이 명확하게 처리가 가능하다고 생각하기 때문입니다. 그렇기에 채넛이 생각했을 때 어떤 부분이 null 값 보다 lateinit이 낫다고 판단했는지를 알고 싶네요!

yrsel replied:
> lateinit 키워드를 사용했을 때, 초기화 되지 않았을 때의 문제 상황과 초기화는 됐지만 Memory Kill이 발생하게 된 경우에 어떻게 대응할 수 있을 지 ?

초기화 되지 않은 경우를 피하기 위해 `isInitialized` 메서드로 사용하기 전에 초기화 여부를 판단하는 방법도 있을 것 같습니다. 하지만, 보일러 플레이트 코드가 많이 늘어날 수 있을 것 같다는 생각이 듭니다.
다른 대안으로 `lateinit`이 아닌 `nullable`한 타입을 사용하는 것이 방법일 수 있을 것 같습니다 !

Memory Kill 에 대한 상황이라면 데이터가 유실된 상황일까요 ?
어떤 상황, 의미인지 정확하게 파악하지 못해서 어떻게 대응해야 할 지 모르겠습니다..😅

>  어떤 부분이 `null` 값 보다 `lateinit`이 낫다고 판단했는지를 알고 싶네요!

저는 NPE 발생 가능성을 만들지 않는 것이 가장 좋다고 생각해서 `lateinit`을 사용하는 것이 낫다고 생각했습니다.
`lateinit`은 초기화 시점만을 주의깊게 판단하면 되지만 `nullable`한 타입은 사용되는 모든 곳에서 `null`에 대한 의심을 해야된다고 생각했습니다.

예시 들어주신 부분의 경우 `null`이라는 컴퓨터 세계에서 통용되는 약속을 따르는 것이 우리끼리의 약속을 통해서 만들어진 값보다 명확한 의미를 전달할 수 있다는 점이 와닿았고 이해하기 쉬웠습니다, 감사합니다 :)

lee-ji-hoon replied:
> Memory Kill 에 대한 상황이라면 데이터가 유실된 상황일까요 ?
어떤 상황, 의미인지 정확하게 파악하지 못해서 어떻게 대응해야 할 지 모르겠습니다..😅

우선 Memory Kill이 그리 흔한 상황은 아니긴 합니다. 
```
1. 장시간 앱을 백그라운드에 뒀다가 돌아오거나
2. adb로 직접 테스트를 한다
...
```
그런데 실제로 Application이 Kill이 났으면 `onCreate`부터 다시 타기에 lateInit은 무조건 초기화가 되어있긴 할 겁니다. 그런데 Activity가 아니라 제가 예시로 넣은 Foo 처럼 전역적으로 lateinit을 하게 되어 있는 경우는 예기치 못한 경우가 발생할 수 있기에 극단적인 예시를 들어드린 것이긴 해서 지금 사실 대응할 것은 없긴 하며, 이런 것도 신경을 써야 한다. 라는 정도로 받아들여주시면 될 것 같습니다.

---

`nullable` 하지 않게 하는 것이 여러 kotlin 책에서도 권장하고 있기는 합니다. 하지만 권장일 뿐 `nullable` 이 더 효과적인 상황은 충분히 있기에 항상 배제하기 보다는 지금처럼 고민을 항상 하길 바랐는데 잘 선택해주신 것 같네요.

> 위 내용과 별개로 `nullable` 과 관련해서 [어렵지만 재밌는 글](https://johngrib.github.io/wiki/pattern/null-object/)이 하나 있어서 추천드리며 나중에 시간 되실 때 읽어보시면 좋을 것 같네요

yrsel replied:
책 엘레강트 오브젝트에서도 null 값 대신 Null Object를 생성하라는 내용을 접했었습니다.
공감은 되었지만 책을 읽을 때 실용적인 측면에서 적용에 어려움이 있지 않을까 생각했었는데 다양한 책에서 Null Object를 제안해주는군요!

좋은 정보 감사합니다 :)

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/BookingResult.kt:
```diff
@@ -0,0 +1,49 @@
+import android.os.Parcelable
+import kotlinx.parcelize.Parcelize
+
+@Parcelize
+data class BookingResult(
+    val title: String,
+    private var _headCount: Int,
+    private var _selectedDate: String,
+    private var _selectedTime: String,
+) : Parcelable {
+    val headCount get() = _headCount
+    val selectedDate get() = _selectedDate
+    val selectedTime get() = _selectedTime
```
> 🙋‍♂️ 질문 3. BookingResult를 immutable 하게 유지하고 싶었지만, 일부 값들이 외부에서 변경되는 상황이 발생해 backing property를 사용해 값을 캡슐화했습니다. 이런 방식이 최선이었을지 고민되고, 더 나은 구조나 대안이 있다면 알고 싶습니다.

내 충분히 가능한 구현이라고 생각은 합니다.

하지만 이 구현에서는 문제가 되는 부분은 여러 Thread에서 이 클래스의 함수에 접근을 해서 변경을 하게 되면 어떻게 될 것인가? 이것을 생각해 보면 다른 방법으로도 지금 구현이 가능해 보이는데요.

가령 val로 두고 copy를 하는 형태로 구현을 변경해 볼 수 있을 것 같은데 해당 구조로 변경을 하려면 많은 것이 변경이 되어야 하는데 레벨 1 때 했던 것을 다시 되짚어 보다 보면 답이 나올 거 같은데 한 번 고민해 보시기 바랍니다.

yrsel replied:
매서드 내부에서 copy 해서 새로운 객체를 반환하는 방식으로 변경하였습니다.

관련 커밋 : 9df7383edb0ab512abf775305084e6855478ad6b

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/Movie.kt:
```diff
@@ -0,0 +1,34 @@
+
+@Parcelize
+data class Movie(
+    @DrawableRes val imageSource: Int,
```
도메인 모델이 이것을 아는게 맞을까요 👀

yrsel replied:
UI의 관심사라고 생각됩니다. 
다만, 외부에 둔다면 어떻게 연결해줄지에 대해 고민이 됩니다. 
크루와 얘기했을 때, 도메인 역할을 하는 model 객체, ui 역할을 하는 ui 모델 객체 2가지로 분리해서 사용하는 방법은 어떨지 이야기를 나눴습니다.

제시해주신 방향이 domainModel, uiModel 2가지로 분리하는 방향이 맞을까요 ??

lee-ji-hoon replied:
네 충분히 가능한 방법입니다. 지금 심지어 `@Parcelize` 를 선택했기에 domain의 model과 ui model 둘로 나누는 것도 좋은 선택으로 보입니다

yrsel replied:
나름대로 domain, ui 모델을 분리해보고 mapper를 통해 변환해주도록 진행했는데 적절하게 사용되었는 지 확인 부탁드립니다 !

관련 커밋  
- ae0eccdcd613d30f6be96e141f3d3d4ca8951299 (Movie 도메인, ui 분리)
- eac356c75c8983d372c0314da7a8874fd1695e03 (BookingResult 도메인, ui 분리)
- da9af3e5b00003689b705c02040fbf110f8af15c ( ui 모델 반환 값 수정 및 상수화 )

lee-ji-hoon replied:
저도 어떤 구조로 변경이 될까 기대가 됐는데, 꽤나 재밌는 구조로 변경이 되었네요.

하나 어색한 부분이 있는데요. Mapper의 역할만 봤을 때 `posterDrawableMapper` 같은 `상태` 를 갖고 있어도 될까?라는 고민을 해보면 어떨까요? 

---

추가로 지금 상황에서는 사실 id 혹은 영화이름을 기반으로 하드코딩된 값을 기반으로 확인하는 게 어떻게 보면 최선으로 볼 수도 있긴 합니다. 주어진 상황 자체가 Poster를 유동적으로 갖고 오기에 적합한 상황은 아니기 때문입니다. 그래도 새로운 시도를 해보았던 점은 좋습니다.

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+        val bookingRunningTime = findViewById<TextView>(R.id.tv_booking_running_time)
+
+        bookingTitle.text = movieData.title
+        val screeningStartDate = formatDate(movieData.screeningStartDate, '.')
+        val screeningEndDate = formatDate(movieData.screeningEndDate, '.')
```
> 🙋‍♂️ 질문 1. 날짜 및 시간 포맷팅 책임의 적절한 위치
이번 미션을 진행하면서 날짜와 시간 포맷팅이 필요한 상황이 있었습니다.
View에서는 String 형태로 데이터를 다루고, 도메인 계층에서는 LocalDate나 LocalTime과 같은 객체를 사용하여 비교를 진행하다 보니, Activity나 도메인 계층에서 포맷 변환 로직이 필요했습니다.
단순히 Util 클래스를 만들어 처리하는 것보다는, 책임에 맞는 객체에 포맷팅 역할을 위임하고 싶은데, 이 경우 포맷팅 책임을 어디에 두는 것이 가장 적절할지 고민이 됩니다.

우선 도메인 객체에서 하게 된다면 UI에 표시하기 위한 데이터를 만들게 되는 상황이니 도메인 객체에 대한 SRP를 어기게 되는 것이 아닐까요? Domain Model에 대한 LocalDate나 LocalTime을 어떻게 표시할지는 View 의 역할이니 그 역할에 맞게끔 책임을 갖고 가면 된다고 생각합니다.

위 내용을 기반으로 채넛이 조금만 고민해보면 어디서 해야 할지 그리고 Util에 두는 것도 충분히 고려해볼만하다고 생각합니다.

lee-ji-hoon replied:
이거에 대해서는 어떤 결론이 나왔을까요 👀

yrsel replied:
커밋 내역 남긴 것으로 착각했습니다 🙈

현재 DateTime 관리하는 util 내부 메서드들을 가독성을 위해 확장함수로 수정하려고 했다가 3, 4단계 얼른 넘어가야 최종 완성할 수 있을 것 같아 머지 후 수정하도록 하겠습니다 🥲

관련 커밋 
10bf66c32d70f61264c3223cb98f85c68884c729 ( 유틸리티 클래스 생성 )
3fa3e3c50a9140734e47913568b8c6ed06e669c3  ( 상수 추가 )
da9af3e5b00003689b705c02040fbf110f8af15c ( 추가 상수화 진행, 내역 중 일부 포함 )

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/MovieAdapter.kt:
```diff
@@ -0,0 +1,74 @@
+import java.time.LocalDate
+import java.time.format.DateTimeFormatter
+
+class MovieAdapter(
+    val context: Context,
+    val movieList: List<Movie>,
+) : BaseAdapter() {
```
Adapter도 Model이라면 context를 알아도 될까요 🤔

yrsel replied:
관련 커밋 : cb27753983aa6aef07189c8ef537dfcf61b76bdb

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/BookingActivity.kt:
```diff
@@ -0,0 +1,242 @@
+package woowacourse.movie
```
Model들을 Package 나눠주신 것 처럼 UI도 한 번 나눠보는 연습을 해보면 좋을 것 같네요.

그렇다고 `move.ui` 에 전부 넣게 되면 나중에 Activity가 더 늘어나면 보기 힘들어지니 패키지도 관심사를 기반으로 나눠보는 연습을 해봅시다

yrsel replied:
관련 커밋 : 5c8f912568f22d1afe3f4609dcc3f2737e5fe44a

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/Booking.kt:
```diff
@@ -0,0 +1,67 @@
+        }
+    }
+
+    private fun formatDate(date: LocalDate): String {
+        return date.format(DateTimeFormatter.ofPattern("yyyy-MM-dd"))
+    }
+
+    private fun formatTime(time: LocalTime): String {
+        return time.format(DateTimeFormatter.ofPattern("kk:mm"))
+    }
```
UI에 맞게끔 format을 해주는 것이 model의 역할일까요 👀

lee-ji-hoon commented on app/src/main/res/layout/activity_booking.xml:
```diff
@@ -0,0 +1,177 @@
+        android:layout_width="match_parent"
+        android:layout_height="match_parent">
+
+        <androidx.constraintlayout.widget.ConstraintLayout
+            android:layout_width="match_parent"
+            android:layout_height="wrap_content">
+
+            <ImageView
+                android:scaleType="centerCrop"
+                android:id="@+id/img_booking_poster"
+                android:layout_width="0dp"
+                android:layout_height="341dp"
+                android:src="@drawable/harry_potter"
+                android:contentDescription="@null"
+                app:layout_constraintEnd_toEndOf="parent"
+                app:layout_constraintStart_toStartOf="parent"
+                app:layout_constraintTop_toTopOf="parent" />
+
+            <androidx.constraintlayout.widget.ConstraintLayout
+                android:layout_width="match_parent"
+                android:layout_height="wrap_content"
+                android:padding="20dp"
+                app:layout_constraintBottom_toBottomOf="parent"
+                app:layout_constraintEnd_toEndOf="parent"
+                app:layout_constraintStart_toStartOf="parent"
+                app:layout_constraintTop_toBottomOf="@id/img_booking_poster">
```
ContraintLayout안에 ConstraintLayout을 안 열어도 가능하지 않았을까 싶은데 이 부분은 시간이 될 때 한 번 도전해보시면 좋을 것 같네요

yrsel replied:
관련 커밋 : 3d06fc918362485790f212a5865721c1cbd92e98

lee-ji-hoon commented on app/src/main/res/values/strings.xml:
```diff
@@ -1,3 +1,16 @@
     <string name="app_name">Movie</string>
+    <string name="screening_date_info">상영일: </string>
+    <string name="running_time_info">러닝타임: </string>
+    <string name="screening_info">영화 상영 시작 시간 15분 전까지\n
+        취소가 가능합니다.</string>
```
15분이라는 것은 비즈니스적인 것이 아닐까요? 만약 이벤트성으로 2배 늘려서 30분이 된다면 strings가 하나 더 늘아나게 됩니다

yrsel replied:
관련 커밋 : 84d953ad18d0ccde2b31b37b6ca6ce2419570b96

lee-ji-hoon reviewed (changes requested):
안녕하세요 채넛 저도 레벨2 첫 미션 잘 부탁드립니다.

우선 채넛이 고민을 하고 있는 부분을 위주로 리뷰를 남겨놨으며 우선 해당 리뷰에 대해서 고민을 해보고 채넛의 의견 혹은 리뷰를 반영해보셨으면 하네요!

고민이 됐던 부분을 되게 잘 적어주셔서 리뷰하기 편했고 어떠한 부분을 고민하셨는지 집중해서 리뷰하기 좋았습니다.

진행을 하면서 어렵거나 이해가 안 되는 부분이 있다면 언제든 편하게 채널 혹은 dm으로 연락 부탁드립니다.

yrsel committed:
style: given-when-then 구조 설정

yrsel committed:
style: 테스트 프로퍼티 선언 및 불필요한 코드 제거

yrsel committed:
refactor: 실패하는 테스트 코드 수정

yrsel committed:
test: 화면 회전 시 데이터 유지 테스트 구현

yrsel committed:
rename: 메서드명 수정

- setupScreeningTimeSpinner -> setUpScreeningTimeSpinner

yrsel committed:
feat: IntentCompat 내부에서 API 버전 매핑하도록 구현

yrsel committed:
rename: API 버전에 따라 Insets 조정하는 함수명 변경

- setUpUi() -> applySystemBarInsets()

yrsel committed:
feat: 키 값 관심사 분리

- Extra -> Intent 의 Key 값들
- SavedState -> SavedInstanceState 의 Key 값들

yrsel committed:
refactor: 여러 영화 데이터 받을 수 있도록 수정

yrsel committed:
refactor: 어댑터 생성자에서 Context 제거

yrsel committed:
feat: 날짜, 시간 유틸리티 클래스 생성

yrsel committed:
refactor: 예약 결과 불변 객체로 변경

yrsel committed:
refactor: 누락된 key값 수정

yrsel committed:
style: 패키지 구조 변경

yrsel committed:
refactor: 리스트뷰 뷰홀더 구현

yrsel committed:
refactor: ConstraintLayout 한 겹 제거

yrsel committed:
refactor: 취소 가능 시간에 대한 정보 위치 수정

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/util/Keys.kt:
```diff
@@ -0,0 +1,15 @@
+package woowacourse.movie.util
+
+object Keys {
+    object Extra {
+        const val LOADED_MOVIE_ITEMS = "extra_loaded_movie_items"
+        const val SELECTED_MOVIE_ITEM = "extra_selected_movie_item"
+        const val BOOKING_RESULT = "extra_booking_result"
+    }
+
+    object SavedState {
+        const val BOOKING_HEAD_COUNT = "savedstate_booking_head_count"
+        const val BOOKING_SCREENING_DATE = "savedstate_booking_screening_date"
+        const val BOOKING_SCREENING_TIME = "savedstate_booking_screening_time"
+    }
+}
```
Key를 별도로 빼는 것도 방법인데 다른 방법도 하나 있습니다.

Key가 직접 사용이 되는 Class에 동반 객체로 넣는 것입니다.

각 방법마다 고려를 해야 할 것들이 있는데, 우선 지금 현재 구현하신 방법에서는 키가 100개가 넘어가는 상황이 왔을 때를 고민해 보면 이 object는 엄청나게 비대해지고 유지보수가 안 되는 key들이 생기게 되는데, 만약 Key를 직접 사용하는 Class의 동반 객체로 넣으면 어떤 단점이 있을까요?

yrsel replied:
이 방법도 고려했으나, 지금 구조와 동일하게 Key를 public하게 선언해줘야하다보니 어디서든 접근할 수 있어 object 클래스로 따로 분리하는 방향이 더 괜찮다고 생각되었습니다.
또한, class의 동반 객체로 선언하게 된다면 Key의 이름이 정확히 기억나지 않을 때, 어떤 클래스에 선언 되어있는지 한번 더 생각하고 가져와야하는 단점이 존재한다고 생각합니다.
만약 Key를 클래스의 동반객체로 저장한다면 데이터를 담아주는 쪽에 키를 저장해서 사용할 것 같습니다 !
장단점을 잘 고려해서 Key를 어디에 저장할 지 결정하도록 하겠습니다 🫡

lee-ji-hoon replied:
> Key를 클래스의 동반객체로 저장한다면 데이터를 담아주는 쪽에 키를 저장해서 사용할 것 같습니다 !

네 저는 이것의 이점이 크다고 생각하는데요. 어떠한 `Activity`를 `start` 할 때 어떠한 key를 넘겨야 하는지를 볼 때 `Keys`라는 것과 해당하는 `Activity` 둘 다 보다 보면 놓치는 key가 생기기 쉽기 때문이고, 해당 `Activity` 에서 필요로 하는 key 이기에 갖고 있는 게 저는 어색하다고 생각하지 않기 때문입니다.

우선 채넛은 별도의 object로 두는 것을 선택했기에 이 방법으로 진행을 해보시면서 어떠한 단점이 있는지도 계속 고민해 보시면 좋을 것 같네요.

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/movie/MovieAdapter.kt:
```diff
@@ -0,0 +1,98 @@
+package woowacourse.movie.model.movie
```
여전이 MovieAdapter가 model에 있기에 하나 더 질문을 드려보겠습니다.

```
model이 findViewById로 View를 알아도 되나요?
```

yrsel replied:
이전 커밋에서 생성자 파라미터에서 context 제거하는 것을 제안해주셨을 때, 의도를 단번에 파악하지 못했습니다.
Adapter는 UI에 가까운 것 같아 패키지 이동하였습니다.

관련 커밋 : 73848e4ff113509f2aa179145da02ecb5e840c20

lee-ji-hoon replied:
네 좋습니다. 조금 더 명확하게 말씀을 드리자면 `Adapter가 Model 패키지에 있는 것이 맞는가?` 였습니다

직접 view의 text를 세팅해주고 있는데 이것이 Model의 역할은 아닐 것이기 때문입니다.   
그렇기에 model에 있을 것이라면 context가 없어야 하는 것 아닌가요?로 돌려서 이야기를 해서 채넛이 오해를 하게 됐었네요

추가로 ui 패키지에 들어간다면 context도 갖고 있어도 이제 큰 문제가 없게 됐네요 👍

yrsel replied:
바로 답을 알려주시는 것보다 간접적인 키워드를 제공해주며 직접 고민할 시간을 주신 점 감사합니다 ! :)

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/booking/BookingResult.kt:
```diff
@@ -0,0 +1,44 @@
+
+    companion object {
+        const val CANCELLATION_LIMIT_MINUTES = 15
+        private const val TICKET_PRICE = 13_000
```
지금은 `13,000원` 으로 티켓 가격을 고정을 했지만, `티켓` 이라는 특성상 충분히 가격의 변동이 있을 수 있을 것 같은데요. 저희는 레벨1 에서 이런 상황을 마주한 적이 있기에 잘 고민해보시면 정답이 나올 것 같네요

yrsel replied:
가격에 대한 책임을 갖는 객체를 생성했는데 의도한 방향으로 가고 있는 지 궁금합니다.

관련 커밋 : 8b3570d3390d50d5ae0f9f14f381fbe3d951ed45

lee-ji-hoon replied:
제가 조금 더 의도하고자 했던 방향은 `로또` 미션 때를 떠올려보면 좋을 것 같네요.

`로또 머신` 이 추가될 것을 대비해서 어떤 방향으로 했고, 막상 `로또 머신` 이 추가가 됐을 때는 어떤 방향으로 구현을 했는지를 다시 떠올려보면 채넛이 금방 이해할 것이라고 믿습니다

lee-ji-hoon commented on app/src/androidTest/java/woowacourse/movie/BookingActivityTest.kt:
```diff
@@ -0,0 +1,315 @@
+package woowacourse.movie
```
테스트의 대상이 되는 클래스들의 패키지 구조가 수정이 되었다면 Test 대상들도 그에 걸맞게 수정이 되어야 합니다.

> UnitTest 부분도 마찬가지입니다.

yrsel replied:
관련 커밋 : dd38e5a67ff2f4b9522e9a520e650be3e2171a52

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/booking/Booking.kt:
```diff
@@ -0,0 +1,64 @@
+        private const val SCREENING_TIMES = 8
+        private const val WEEKDAY_START_TIME = 9
+        private const val WEEKEND_START_TIME = 10
+        private const val DATE_FORMAT = "yyyy-MM-dd"
+        private const val TIME_FORMAT = "kk:mm"
```
전반적으로 이런 값들이 계속 재사용이 되고 있는데 이게 조금 더 어울리는 Util이 하나 보이네요

yrsel replied:
DateTimeUtil 내부로 이동하였습니다 

관련 커밋 : 3fa3e3c50a9140734e47913568b8c6ed06e669c3

lee-ji-hoon commented on app/src/main/java/woowacourse/movie/model/movie/MovieAdapter.kt:
```diff
@@ -0,0 +1,98 @@
+            registerReserveOnClickListener(movie)
+        }
+
+        private fun registerReserveOnClickListener(movie: Movie) {
+            reserveButton.setOnClickListener {
+                val context = view.context
+                val intent =
+                    Intent(context, BookingActivity::class.java).apply {
+                        this.putExtra(Keys.Extra.SELECTED_MOVIE_ITEM, movie)
+                    }
+                context.startActivity(intent)
+            }
+        }
+    }
```
Adapter의 경우 여러 화면에서 재사용이 될 수 있는데요. 그런 재사용적인 측면에서 만약 A라는 화면에서는 예약 확정 페이지가 아닌 이벤트 페이지로 보내주려고 한다면 Adapter가 새로 만들어져야 하는 상황인데요. 어떻게 바꿔 볼 수 있을까요?

yrsel replied:
오늘까지는 제출해야 3, 4단계를 마무리할 수 있을 것 같아 제출 후 3, 4단계 반영하며 같이 반영해보겠습니다. ! 🥲

lee-ji-hoon reviewed (changes requested):
채넛 안녕하세요! 반영해주신 부분 잘 확인했으며, 커밋 내용도 남겨주셔서 감사합니다~!

3,4 단계로 넘어가기 전에 채넛이 확인해봤으면 하는 부분과 고민 중인 내용들에 대해서 리뷰를 남겨놨습니다.

yrsel committed:
style: 패키지 구조 변경

yrsel committed:
rename: 메인 액티비티 이름 수정

- MainActivity.kt -> MovieActivity.kt
- MainActivityTest.kt -> MovieActivityTest.kt

yrsel committed:
style: 어댑터 패키지 이동

- model -> ui 패키지로 이동

yrsel committed:
style: MovieActivityTest 패키지 수정

yrsel committed:
refactor: 날짜, 시간 상수 선언 추가

yrsel committed:
refactor: Movie 도메인,ui 모델 분리

yrsel committed:
style: movieUiModel 타입 명시 및 변수명 수정

yrsel committed:
refactor: BookingResult 도메인,ui 모델 분리

yrsel committed:
refactor: UI 모델 수정 및 구분자 상수화

yrsel committed:
fix: 실패하는 테스트 코드 정상동작하도록 수정

yrsel committed:
fix: xml 확인용 데이터 tools 로 변경

yrsel committed:
refactor: 가격 외부에서 계산할 수 있도록 수정

yrsel committed:
refactor: restore lifecycle 사용

yrsel commented:
안녕하세요, 코니 ☘️
도메인 모델과 ui 모델을 분리하는 과정에서 고민이 많이 생겨 생각보다 시간이 오래 걸렸습니다.
이번에 반영 못한 리뷰들을 3, 4단계 미션을 진행하며 반영하겠습니다 ! 
현재 BookingActivity에 너무 많은 로직들이 담겨있어 스피너를 분리할 계획이고 그 외 리팩터링 필요한 부분도 3, 4단계를 진행할 때 같이 반영하도록 하겠습니다.
1, 2단계 리뷰해주셔서 감사합니다 :)

lee-ji-hoon reviewed (approved):
채넛 안녕하세요! 1,2단계는 여기서 마무리하고자 합니다.

남겨주신 질문에 대한 답변은 달아뒀으며, 해당 리뷰에 대해서 고민이 길어진다면 3,4단계를 진행하면서 고민해 보다가 이해가 안 된다면 dm 혹은 채널에 연락 부탁드리겠습니다.

3,4단계도 파이팅입니다~!
